### PR TITLE
Discord Integration

### PR DESCRIPTION
**Pull request by CalumTowers**
- Implements a number of front-end pages to support user registration to the VATSIM UK Discord server
- Allows a Discord account to be linked to an Account in Core through Discord's OAuth2 server
- By using a command (`php artisan discord:manager`), complete a number of actions using the Discord API; keep everyone's nickname up to date and grant them the relevant roles as per `discord_roles` (linked to `mship_permission`)
- When the user unlinks their Discord account, it will kick them from the server (Discord removes all roles on their side so that if they rejoin they will need to link their account again)

### CODE REVIEW

File: config/services.php [hunk 1]
```diff
@@ -40,12 +40,13 @@
...
-        'client_id' => env('DISCORD_CLIENT'),
-        'client_secret' => env('DISCORD_SECRET'),
-        'redirect_uri' => env('DISCORD_REDIRECT_URI'),
+        'invite_url' => env('DISCORD_INVITE_URL', null),
+        'guild_id' => env('DISCORD_GUILD_ID', null),
+        'token' => env('DISCORD_TOKEN', null),
+        'client_id' => env('DISCORD_CLIENT', null),
+        'client_secret' => env('DISCORD_SECRET', null),
+        'redirect_uri' => env('DISCORD_REDIRECT_URI', null),
+        'base_discord_uri' => env('DISCORD_API_BASE', 'https://discordapp.com/api/v6')
```
Comment (CalumTowers) [inline config/services.php:49]: I had this URI hard coded as it was a way of showing that the Library for Discord had been written for V6 of the API. Might be worth just commenting that into the library if we are extracting the URI to a config item

File: public/vendor/telescope/app-dark.css [hunk 2]
```diff
@@ -1,8 +1,8 @@
 .vjs-tree{font-family:Monaco,Menlo,Consolas,Bitstream Vera Sans Mono,monospace!important}.vjs-tree .vjs-tree__content{border-left:1px dotted hsla(0,0%,80%,.28)!important}.vjs-tree .vjs-tree__node{cursor:pointer}.vjs-tree .vjs-tree__node:hover{color:#20a0ff}.vjs-tree .vjs-checkbox{position:absolute;left:-30px}.vjs-tree .vjs-value__boolean,.vjs-tree .vjs-value__null,.vjs-tree .vjs-value__number{color:#a291f5!important}.vjs-tree .vjs-value__string{color:#dacb4d!important}.hljs-addition,.hljs-keyword,.hljs-selector-tag{color:#8bd72f}.hljs-doctag,.hljs-meta .hljs-meta-string,.hljs-regexp,.hljs-string{color:#dacb4d}.hljs-literal,.hljs-number{color:#a291f5!important}
 
 /*!
- * Bootstrap v4.3.1 (https://getbootstrap.com/)
```
Comment (atoff) [inline public/vendor/telescope/app-dark.css:4]: Not sure that these `public/vendor` files should be tracked in git?

File: resources/views/discord/new.blade.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                            Whether you've used Discord before or not, clicking the button below will take you to a page to either sign in or register with Discord.
+                        </p>
+                        <br />
+                        <p>
+
+                            @if(!$_account->discord_id)
+                                <a href="{{ route('discord.invite') }}" style="text-decoration: none;" target="_blank">
+                                    <button class="btn btn-primary center-block"><em class="fab fa-discord"></em> &thinsp; Join Discord Server</button>
+                                </a>
+                            @endif
```
Comment (atoff) [inline resources/views/discord/new.blade.php:45]: `@else`

File: resources/views/discord/new.blade.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                        <p>
+                            This allows us to assign you the relevant permissions on our Discord server.
+                        </p>
+                        <br />
+                        <p>
+                            @if(!$_account->discord_id)
+                                <a href="{{ route('discord.create') }}" style="text-decoration: none;">
+                                    <button class="btn btn-primary center-block"><em class="fab fa-discord"></em> &thinsp; Link Discord Account</button>
+                                </a>
+                            @endif
```
Comment (atoff) [inline resources/views/discord/new.blade.php:74]: `@else`

File: app/Models/Discord/DiscordRole.php [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+namespace App\Models\Discord;
+
+use Illuminate\Database\Eloquent\Model;
+use Spatie\Permission\Models\Permission;
+
+class DiscordRole extends Model
+{
+    public $timestamps = false;
+
+    public function role()
```
Comment (atoff) [inline app/Models/Discord/DiscordRole.php:12]: Should this be `permission`?

File: app/Models/Discord/DiscordRole.php [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace App\Models\Discord;
+
+use Illuminate\Database\Eloquent\Model;
+use Spatie\Permission\Models\Permission;
+
+class DiscordRole extends Model
+{
+    public $timestamps = false;
+
```
Comment (atoff) [inline app/Models/Discord/DiscordRole.php:11]:
```
protected $casts = [
 'permission_id' => 'int'
];
```

File: app/Console/Commands/ExternalServices/ManageDiscord.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        });
+    }
+
+    protected function removeRoles()
+    {
+        $account = $this->account;
+        $discord = $this->discord;
+
+        DiscordRole::all()->filter(function ($value) use ($account) {
+            return ! $account->hasPermissionTo((int) $value['permission_id']);
```
Comment (atoff) [inline app/Console/Commands/ExternalServices/ManageDiscord.php:103]: Recommend use the casts in the model (See comment on DiscordRole model)

File: database/migrations/2020_04_30_215425_add_discord_id_column.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * Run the migrations.t.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('mship_account', function (Blueprint $table) {
+            $table->string('discord_id', 30)->unique()->after('slack_id')->nullable();
```
Comment (atoff) [inline database/migrations/2020_04_30_215425_add_discord_id_column.php:17]: Do we also want a migration to remove slack_id's?

Comment (CalumTowers) [reply database/migrations/2020_04_30_215425_add_discord_id_column.php:17]: Not for the time being, we'll remove it later

File: public/vendor/telescope/app-dark.css (hunk 2 above)
Comment (CalumTowers) [reply public/vendor/telescope/app-dark.css:4]: They are vendor files that we have published to our local repo that we can then modify, so I think this is fine

File: app/Models/Mship/Account.php [hunk 9]
```diff
@@ -43,6 +41,7 @@
  *
  * @property int $id
  * @property string|null $slack_id
```
[the remark this answers is not in the document]
Comment (CalumTowers) [reply app/Models/Mship/Account.php:43]: Soon, but not yet

File: app/Console/Commands/Development/TestEmails.php [hunk 10]
```diff
@@ -134,7 +133,6 @@ public function handle()
         $testAccount->notify(new BanRepealed($testBan));
         $testAccount->notify(new EmailVerification($testEmail, $testTokenEmailVerify));
         $testAccount->notify(new FeedbackReceived($testFeedback));
-        $testAccount->notify(new SlackInvitation());
```
[the remark this answers is not in the document]
Comment (CalumTowers) [reply app/Console/Commands/Development/TestEmails.php:137]: We'll sort that later - although this command missed a lot of emails from our system

File: app/Console/Commands/ExternalServices/ManageDiscord.php [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    /**
+     * Create a new command instance.
+     *
+     * @return void
+     */
+    public function __construct()
+    {
+        parent::__construct();
+
+        $this->discord = app()->make(Discord::class);
```
[the remark this answers is not in the document]
Comment (CalumTowers) [reply app/Console/Commands/ExternalServices/ManageDiscord.php:43]: This is a line that @AxonC added so I'll tag him for thoughts

File: app/Console/Commands/ExternalServices/ManageDiscord.php [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        $discordUsers = $this->getUsers();
+
+        if (! $discordUsers) {
+            $this->error('No users found.');
+            exit();
+        }
+
+        foreach ($discordUsers as $account) {
+            $this->account = $account;
+            $this->grantRoles();
```
[the remark this answers is not in the document]
Comment (CalumTowers) [reply app/Console/Commands/ExternalServices/ManageDiscord.php:62]: We could do, but then we end up passing the account to all of the different methods, this feels cleaner as we reference the account from one place.
Either way, I think it is readable?

File: app/Console/Commands/ExternalServices/ManageDiscord.php [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        $this->info($discordUsers->count().' user(s) updated on Discord.');
+        Log::debug($discordUsers->count().' user(s) updated on Discord.');
+    }
+
+    protected function getUsers()
+    {
+        if ($this->option('force')) {
+            return Account::where('id', $this->option('force'))->where('discord_id', '!=', null)->get();
```
[the remark this answers is not in the document]
Comment (CalumTowers) [reply app/Console/Commands/ExternalServices/ManageDiscord.php:74]: I was going to do this but my view is that your code snippet is a lot harder to read vs mine. Imagine you didn't know what it was doing, much easier to follow (imo) when it is all in front of you.

File: app/Console/Commands/ExternalServices/ManageDiscord.php [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+    protected function assignNickname()
+    {
+        $this->discord->setNickname($this->account, $this->account->name);
+    }
+
+    protected function grantRoles()
+    {
+        $account = $this->account;
+        $discord = $this->discord;
```
[the remark this answers is not in the document]
Comment (CalumTowers) [reply app/Console/Commands/ExternalServices/ManageDiscord.php:88]: We reference it twice, in the function itself and in the use params.
If you use $this->discord and $this->account in both the use and the statement it looks a tad messy

File: resources/views/discord/new.blade.php (hunk 3 above)
Comment (CalumTowers) [reply resources/views/discord/new.blade.php:45]: b8985c8da54d63fccefb9850e0d980be96da8883

File: app/Console/Commands/ExternalServices/ManageDiscord.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (atoff) [reply app/Console/Commands/ExternalServices/ManageDiscord.php:62]: I think its ok, but assigning the account into the class private variable doesn't really make sense in the context of the class and isn't generally done

File: app/Console/Commands/ExternalServices/ManageDiscord.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (AxonC) [reply app/Console/Commands/ExternalServices/ManageDiscord.php:43]: Yeah, that works, typehint in the constructor and assign that. Didn't think commands were automatically resolved (diesn't say it in the docs mind.

File: resources/views/discord/new.blade.php (hunk 4 above)
Comment (CalumTowers) [reply resources/views/discord/new.blade.php]: b8985c8

File: app/Console/Commands/ExternalServices/ManageDiscord.php (hunk 7 above)
Comment (CalumTowers) [reply app/Console/Commands/ExternalServices/ManageDiscord.php]: 2a66516

File: app/Models/Discord/DiscordRole.php (hunk 6 above)
Comment (CalumTowers) [reply app/Models/Discord/DiscordRole.php:11]: 2a66516

File: app/Models/Discord/DiscordRole.php (hunk 5 above)
Comment (CalumTowers) [reply app/Models/Discord/DiscordRole.php]: Removed in c2d4c3c as the relationship is not used anywhere

File: app/Console/Commands/ExternalServices/ManageDiscord.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (CalumTowers) [reply app/Console/Commands/ExternalServices/ManageDiscord.php]: 26348d8

File: app/Console/Commands/ExternalServices/ManageDiscord.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (atoff) [reply app/Console/Commands/ExternalServices/ManageDiscord.php:88]: If you use `$this->discord`, you shouldn't have to pass in `$discord` in the use? (I think, can't remember of the top of my head)

**Pull request merged**